### PR TITLE
Fix async transaction problem when async query already exist.

### LINKED ISSUE #1893: onSuccess might not be called with the right Realm version for async transaction
Now the `onSuccess` relies on the `REALM_CHANGED` triggers the `advance_read` first. If it is running with a unfinished async query, the `REALM_CHANGED` might be swallowed by `updateAsyncQueries` and the Realm doesn't make a `andvance_read`.

In above case, the next message in the handler queue is to run `onSuccess`, but it the function will be called in a wrong earlier Realm version which the changes in the async transaction is not in.

--------------------- Update Jan. 25th 2016 -------------------------
http://stackoverflow.com/questions/34938152/callback-never-be-invoked-when-i-excute-a-asynchronous-transaction-with-realm

See above question on SO. We need to discuss of the user case of `onSuccess`. Currently if the Realm instance is closed, then the `onSuccess` won't be called depends on the timing. But it might not be what user expect.

Shall we just let `onSuccess` be delivered whenever the transaction is done, and document it clearly that in the `onSuccess`, the Realm might not be the latest version?

### PR DESCRIPTION
**Pull request by dalinaum**
Fix #1893 It imports #2045 for test.

### CODE REVIEW

**Comment by dalinaum** [comment]
Please review this @realm/java

File: realm/realm-library/src/main/java/io/realm/BaseRealm.java [hunk 1]
```diff
@@ -66,6 +67,7 @@
     RealmSchema schema;
     Handler handler;
     HandlerController handlerController;
+    List<Runnable> asyncTransactionCallbacks = new ArrayList<>();
```
Comment (zaki50) [inline realm/realm-library/src/main/java/io/realm/BaseRealm.java]: we should avoid diamond operator

File: realm/realm-library/src/main/java/io/realm/HandlerController.java [hunk 2]
```diff
@@ -509,6 +509,13 @@ private void completedAsyncQueriesUpdate(QueryUpdateTask.Result result) {
 
             updateAsyncQueriesTask = null;
         }
+
+        if (realm.asyncTransactionCallbacks.size() > 0) {
```
Comment (zaki50) [inline realm/realm-library/src/main/java/io/realm/HandlerController.java]: Use `!isEmpty()` instead of `size() > 0`

Comment (dalinaum) [inline realm/realm-library/src/main/java/io/realm/HandlerController.java]: Updated. Thanks.

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 3]
```diff
@@ -1138,87 +1138,19 @@ public void executeTransaction(Transaction transaction) {
      */
     @Deprecated
     public RealmAsyncTask executeTransaction(final Transaction transaction, final Transaction.Callback callback) {
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: Perhaps just delete this instead? It has been deprecated for a while now?

File: realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java [hunk 4]
```diff
@@ -1947,6 +1947,36 @@ public void onChange() {
...
     }
 
+    @Test
+    @RunTestInLooperThread
+    public void testAsyncTransactionWorksWithAsyncQuery() {
+        RealmResults<AllTypes> results = looperThread.realm.where(AllTypes.class).findAllAsync();
+        results.addChangeListener(new RealmChangeListener() {
+            @Override
+            public void onChange() {
+                //assertEquals(looperThread.realm.where(AllTypes.class).count(), 1);
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java]: If we don't use this listener, perhaps just remove it?

File: realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java [hunk 5]
```diff
@@ -1947,6 +1947,36 @@ public void onChange() {
         result.load();
     }
 
+    @Test
+    @RunTestInLooperThread
+    public void testAsyncTransactionWorksWithAsyncQuery() {
+        RealmResults<AllTypes> results = looperThread.realm.where(AllTypes.class).findAllAsync();
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java]: A pattern we have used elsewhere is:
```
Realm realm = looperThread.realm;
```

It makes it slightly more readable, and it is easier to replace the Realm instance if we need to.

File: realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java [hunk 6]
```diff
@@ -1947,6 +1947,36 @@ public void onChange() {
         result.load();
     }
 
+    @Test
+    @RunTestInLooperThread
+    public void testAsyncTransactionWorksWithAsyncQuery() {
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java]: I could use a comment about what this unit test is for. You are doing nothing with the `findAllAsync` result as far as I can tell?

**Comment by cmelchior** [comment]
This PR only seem to solve half the problem? e.g the problem described here http://stackoverflow.com/questions/34938152/callback-never-be-invoked-when-i-excute-a-asynchronous-transaction-with-realm is not addressed

Looking at it more closely it looks to me like #1893 and #2045 is two different issues? I'm fine with this PR just solving #2045 but we shouldn't close #1893 in that case. Thoughts @beeender ?

**Comment by dalinaum** [comment]
Sorry. I almost forgot that #1893 are actually two issues. I will add logging a warning tonight.

**Comment by dalinaum** [comment]
Use #2780 instead.

**Pull request closed without merging**